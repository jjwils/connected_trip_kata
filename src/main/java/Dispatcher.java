public class Dispatcher {
    private final Taxis taxis;

    public Dispatcher(Taxis allTheTaxis) {

        this.taxis = allTheTaxis;
    }

    public void dispatch(String command) {
       taxis.move("R");
    }
}
