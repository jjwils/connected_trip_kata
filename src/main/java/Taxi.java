import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

public class Taxi {

    private final String taxi;
    private final String location;

    private int xgrid;

    private int ygrid;

    private int cellometer;



    public String getTaxiName() {
        return taxi;
    }
    public Taxi(String taxi, String location) {
        this.taxi = taxi;
        this.location = location;
        this.xgrid = Integer.parseInt(location.split(",")[0]);
        this.ygrid = Integer.parseInt(location.split(",")[1]);
    }

    public String gridref() {
        return "" + this.xgrid + "," + this.ygrid;
    }

    public void move(String move, HashMap<String, Taxi> taxis) {
        boolean canIMove = true;

        for(Map.Entry<String, Taxi> entry : taxis.entrySet()) {
            String taxiName = entry.getKey();
            String otherGridref = entry.getValue().gridref();


            if(!taxiName.equals(this.getTaxiName()) &&
                    this.gridref().equals(otherGridref)
                    ) {
                canIMove = false;
            }
        }

        if(canIMove) {
            actuallyMove(move);
        }
    }

    public void actuallyMove(String command) {
            move(command);
    }

    private void move(String move) {
        move.chars().forEach(movement->{
            cellometer++;
            if (movement == 'U') {
                ygrid--;
            } else if(movement == 'D') {
                ygrid++;
            } else if(movement == 'L') {
                xgrid--;
            } else if(movement == 'R') {
                xgrid++;
            }
        });
    }

    public int cellometer() {
        return cellometer;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Taxi taxi1 = (Taxi) o;
        return Objects.equals(taxi, taxi1.taxi);
    }

    @Override
    public int hashCode() {
        return Objects.hash(taxi);
    }
}
