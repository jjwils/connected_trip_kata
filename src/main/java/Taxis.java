import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

public class Taxis {
    private final List<Taxi> taxiList;

    public Taxis(List<Taxi> taxiList) {

        this.taxiList = taxiList;
    }

    public void move(String command) {

        Taxi topTaxi = getTopTaxi(taxiList);

        int gridsInUse = numberOfGridsInUse(taxiList);

        if (gridsInUse == 3) {
            taxiList.forEach(taxi -> taxi.actuallyMove(command));
        }

        else {

            //lets keep a handle on where the top taxi currently is
            String topTaxiGridref = topTaxi.gridref();

            //top taxi always gets to move regardless
            topTaxi.actuallyMove(command);
            taxiList.remove(topTaxi);

            //with the top taxi no longer counted then if we have two everyone can move
            //unless they were in the same grid as the top taxi
            gridsInUse = numberOfGridsInUse(taxiList);

            if (gridsInUse == 2){
                taxiList.forEach(taxi -> {
                    if (!taxi.gridref().equals(topTaxiGridref))
                    taxi.actuallyMove(command);

                });
            }
            //otherwise the remaining two must be in the same grid so only allow the new toptaxi to move
            //again only if they were not in the same grid as the original top taxi
            else{
                Taxi newTopTaxi = getTopTaxi(taxiList);

                if (!newTopTaxi.gridref().equals(topTaxiGridref))
                {
                    newTopTaxi.actuallyMove(command);
                }
            }

        }
    }

    private int numberOfGridsInUse(List<Taxi> taxis) {
        return taxis.stream().map(taxi -> taxi.gridref()).collect(Collectors.toSet())
                .size();
    }

    private Taxi getTopTaxi(List<Taxi> taxis) {
        return taxis.stream()
                .min(Comparator.comparing(taxi -> taxi.getTaxiName()))
                .orElse(null);
    }

    public List<Taxi> taxis() {
        return taxiList;
    }
}
