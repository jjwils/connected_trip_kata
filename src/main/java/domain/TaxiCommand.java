package domain;

public record TaxiCommand(Taxi taxi, String command) {
}