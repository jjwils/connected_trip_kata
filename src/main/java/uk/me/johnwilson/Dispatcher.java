package uk.me.johnwilson;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

public class Dispatcher {
    private final Taxis taxis;

    public Dispatcher(Taxis allTheTaxis) {
        this.taxis = allTheTaxis;
    }


    public void dispatch(List<TaxiCommand> taxiCommands) {
        int maxLength = taxiCommands.stream()
                .mapToInt(taxiCommand -> taxiCommand.command().length())
                .max()
                .orElse(0);

        for (int i = 0; i < maxLength; i++) {
            assessNextSteps(taxis.taxis());
            for (TaxiCommand taxiCommand : taxiCommands) {
                String commandString = taxiCommand.command();
                if (i < commandString.length()) {
                    char commandChar = commandString.charAt(i);

                    taxiCommand.taxi().move(String.valueOf(commandChar));
                }
            }
        }
    }

    public void dispatch(TaxiCommand taxiCommand) {
        taxiCommand.taxi().actuallyMove(taxiCommand.command());

    }

    public void assessNextSteps(List<Taxi> taxisInput) {
        //assume cant move for all to start
        taxisInput.forEach(taxi -> taxi.canMoveOnNextStep(false));

        //take a copy of the list
        List<Taxi> taxis = new ArrayList<>(taxisInput);

        //highest priority taxi can always move regardless
        Taxi highestPriorityTaxi = highestPriorityTaxi(taxis);
        highestPriorityTaxi.canMoveOnNextStep(true);

        //if all the taxis are in different grids everyone can move
        if (allTaxisInDifferentGrids(taxis)) {
                taxis.forEach(taxi -> taxi.canMoveOnNextStep(true));
        }

        else {
            //takeaway the highest priority taxi
            taxis.remove(highestPriorityTaxi);

            List<Taxi> taxisToRemove = new ArrayList<>();
            //takeaway any taxi that is in the same grid as the hpt
            taxis.forEach(taxi -> {
                if (taxi.gridref().equals(highestPriorityTaxi.gridref())){
                    taxisToRemove.add(taxi);
                }
            });

            taxis.removeAll(taxisToRemove);

            //any taxis left call ourselves recursively
            if (!taxis.isEmpty()){
                assessNextSteps(taxis);
            }

        }


    }

    private boolean allTaxisInDifferentGrids(List taxis) {
        int gridsInUse = ((List<Taxi>) taxis).stream().map(taxi -> taxi.gridref()).collect(Collectors.toSet())
                .size();
        return gridsInUse == taxis.size();
    }

    private Taxi highestPriorityTaxi(List<Taxi> taxis) {
        return taxis.stream()
                .min(Comparator.comparing(taxi -> taxi.getTaxiName()))
                .orElse(null);
    }
}

