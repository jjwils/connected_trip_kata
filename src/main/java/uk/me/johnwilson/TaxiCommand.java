package uk.me.johnwilson;

public record TaxiCommand(Taxi taxi, String command) {
}