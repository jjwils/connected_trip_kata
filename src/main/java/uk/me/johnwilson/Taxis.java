package uk.me.johnwilson;

import java.util.HashMap;
import java.util.List;

public class Taxis {
    private  List<Taxi> taxis;

    private  HashMap<String,Taxi> taxiLocations;

    public Taxis(HashMap<String, Taxi> taxiLocations){
        this.taxiLocations = taxiLocations;

    }

    public Taxis(List<Taxi> taxiList) {
        this.taxis = taxiList;
    }


    public List<Taxi> taxis() {
        return taxis;
    }


    public String taxiLocation(String taxiName) {
        return taxiLocations.get(taxiName).gridref();
    }
}
