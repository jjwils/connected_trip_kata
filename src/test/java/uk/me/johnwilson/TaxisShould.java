package uk.me.johnwilson;

import uk.me.johnwilson.Dispatcher;
import uk.me.johnwilson.Taxi;
import uk.me.johnwilson.TaxiCommand;
import uk.me.johnwilson.Taxis;
import org.approvaltests.Approvals;
import org.approvaltests.reporters.Junit5Reporter;
import org.approvaltests.reporters.UseReporter;
import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;

@UseReporter(Junit5Reporter.class)
public class TaxisShould {


    @Test
    public void start_in_correct_grid()
    {
        Taxi taxiA = new Taxi("A","0,0");
        Taxi taxiB = new Taxi("B","0,1");
        Taxi taxiC = new Taxi("C","2,3");

        List<Taxi> listOfTaxis = new ArrayList<>();
        listOfTaxis.add(taxiA);
        listOfTaxis.add(taxiB);
        listOfTaxis.add(taxiC);


        Approvals.verify(TaxiPrinter.plotTaxis(listOfTaxis));


    }


    @Test
    public void all_move_right_one_grid()
    {
        Taxi taxiA = new Taxi("A","1,0");
        Taxi taxiB = new Taxi("B","1,1");
        Taxi taxiC = new Taxi("C","2,0");

        List<Taxi> listOfTaxis = new ArrayList<>();
        listOfTaxis.add(taxiA);
        listOfTaxis.add(taxiB);
        listOfTaxis.add(taxiC);

        Taxis taxis = new Taxis(listOfTaxis);
        Dispatcher dispatcher = new Dispatcher(taxis);


        List<TaxiCommand> taxiCommands = new ArrayList<>();
        taxiCommands.add(new TaxiCommand(taxiA, "R"));
        taxiCommands.add( new TaxiCommand(taxiB, "R"));
        taxiCommands.add(new TaxiCommand(taxiC, "R"));


        //act
        dispatcher.dispatch(taxiCommands);

        //assert
        Approvals.verify(TaxiPrinter.plotTaxis(listOfTaxis));


    }

    @Test
    public void all_complete_two_non_blocking_moves()
    {
        Taxi taxiA = new Taxi("A","0,0");
        Taxi taxiB = new Taxi("B","2,2");
        Taxi taxiC = new Taxi("C","4,3");

        List<Taxi> listOfTaxis = new ArrayList<>();
        listOfTaxis.add(taxiA);
        listOfTaxis.add(taxiB);
        listOfTaxis.add(taxiC);

        Taxis taxis = new Taxis(listOfTaxis);
        Dispatcher dispatcher = new Dispatcher(taxis);


        List<TaxiCommand> taxiCommands = new ArrayList<>();
        taxiCommands.add(new TaxiCommand(taxiA, "RD"));
        taxiCommands.add( new TaxiCommand(taxiB, "LU"));
        taxiCommands.add(new TaxiCommand(taxiC, "LU"));


        //act
        dispatcher.dispatch(taxiCommands);

        //assert
        Approvals.verify(TaxiPrinter.plotTaxis(listOfTaxis));


    }

    @Test
    public void only_a_move()
    {
        Taxi taxiA = new Taxi("A","0,0");
        Taxi taxiB = new Taxi("B","0,0");
        Taxi taxiC = new Taxi("C","0,0");

        List<Taxi> listOfTaxis = new ArrayList<>();
        listOfTaxis.add(taxiA);
        listOfTaxis.add(taxiB);
        listOfTaxis.add(taxiC);

        Taxis taxis = new Taxis(listOfTaxis);
        Dispatcher dispatcher = new Dispatcher(taxis);


        List<TaxiCommand> taxiCommands = new ArrayList<>();
        taxiCommands.add(new TaxiCommand(taxiA, "R"));
        taxiCommands.add( new TaxiCommand(taxiB, "R"));
        taxiCommands.add(new TaxiCommand(taxiC, "R"));


        //act
        dispatcher.dispatch(taxiCommands);

        //assert
        Approvals.verify(TaxiPrinter.plotTaxis(listOfTaxis));


    }

    @Test
    public void only_a_and_b_move()
    {
        Taxi taxiA = new Taxi("A","0,0");
        Taxi taxiB = new Taxi("B","1,0");
        Taxi taxiC = new Taxi("C","0,0");

        List<Taxi> listOfTaxis = new ArrayList<>();
        listOfTaxis.add(taxiA);
        listOfTaxis.add(taxiB);
        listOfTaxis.add(taxiC);

        Taxis taxis = new Taxis(listOfTaxis);
        Dispatcher dispatcher = new Dispatcher(taxis);


        List<TaxiCommand> taxiCommands = new ArrayList<>();
        taxiCommands.add(new TaxiCommand(taxiA, "R"));
        taxiCommands.add( new TaxiCommand(taxiB, "R"));
        taxiCommands.add(new TaxiCommand(taxiC, "R"));


        //act
        dispatcher.dispatch(taxiCommands);

        //assert
        Approvals.verify(TaxiPrinter.plotTaxis(listOfTaxis));


    }

    @Test
    public void only_a_and_c_move()
    {
        Taxi taxiA = new Taxi("A","0,0");
        Taxi taxiB = new Taxi("B","0,0");
        Taxi taxiC = new Taxi("C","1,0");

        List<Taxi> listOfTaxis = new ArrayList<>();
        listOfTaxis.add(taxiA);
        listOfTaxis.add(taxiB);
        listOfTaxis.add(taxiC);

        Taxis taxis = new Taxis(listOfTaxis);
        Dispatcher dispatcher = new Dispatcher(taxis);


        List<TaxiCommand> taxiCommands = new ArrayList<>();
        taxiCommands.add(new TaxiCommand(taxiA, "R"));
        taxiCommands.add( new TaxiCommand(taxiB, "R"));
        taxiCommands.add(new TaxiCommand(taxiC, "R"));


        //act
        dispatcher.dispatch(taxiCommands);

        //assert
        Approvals.verify(TaxiPrinter.plotTaxis(listOfTaxis));


    }

    @Test
    public void a_block_b()
    {
        Taxi taxiA = new Taxi("A","0,0");
        Taxi taxiB = new Taxi("B","2,0");

        List<Taxi> listOfTaxis = new ArrayList<>();
        listOfTaxis.add(taxiA);
        listOfTaxis.add(taxiB);

        Taxis taxis = new Taxis(listOfTaxis);
        Dispatcher dispatcher = new Dispatcher(taxis);


        List<TaxiCommand> taxiCommands = new ArrayList<>();
        taxiCommands.add(new TaxiCommand(taxiA, "RD"));
        taxiCommands.add( new TaxiCommand(taxiB, "LD"));


        //act
        //dispatcher.assessNextSteps(listOfTaxis);
        dispatcher.dispatch(taxiCommands);

        //assert
        Approvals.verify(TaxiPrinter.plotTaxis(listOfTaxis));


    }





    @Test
    void all_arrive_at_airport_5_2(){
        //arrange

        Taxi taxiA = new Taxi("A","5,3");
        Taxi taxiB = new Taxi("B","2,2");
        Taxi taxiC = new Taxi("C","1,1");

        List<Taxi> listOfTaxis = new ArrayList<>();
        listOfTaxis.add(taxiA);
        listOfTaxis.add(taxiB);
        listOfTaxis.add(taxiC);

        Taxis taxis = new Taxis(listOfTaxis);
        Dispatcher dispatcher = new Dispatcher(taxis);


        List<TaxiCommand> taxiCommands = new ArrayList<>();
        taxiCommands.add(new TaxiCommand(taxiA, "LLUURRD"));
        taxiCommands.add( new TaxiCommand(taxiB, "URRRD"));
        taxiCommands.add(new TaxiCommand(taxiC, "RRRRRD"));

        //act
        dispatcher.dispatch(taxiCommands);

        //assert
        Approvals.verify(TaxiPrinter.plotTaxis(listOfTaxis));

        System.out.println("TaxiA distance travelled: " + taxiA.cellometer());
        System.out.println("TaxiB distance travelled: " + taxiB.cellometer());
        System.out.println("TaxiC distance travelled: " + taxiC.cellometer());

    }

    @Test
    void all_pick_up_and_drop_off(){
        //arrange

        Taxi taxiA = new Taxi("A","5,3");
        Taxi taxiB = new Taxi("B","2,2");
        Taxi taxiC = new Taxi("C","1,1");

        List<Taxi> listOfTaxis = new ArrayList<>();
        listOfTaxis.add(taxiA);
        listOfTaxis.add(taxiB);
        listOfTaxis.add(taxiC);

        Taxis taxis = new Taxis(listOfTaxis);
        Dispatcher dispatcher = new Dispatcher(taxis);


        List<TaxiCommand> taxiCommands = new ArrayList<>();
        taxiCommands.add(new TaxiCommand(taxiA, "LLUURRDULLLLD"));
        taxiCommands.add( new TaxiCommand(taxiB, "URRRDUULLLLD"));
        taxiCommands.add(new TaxiCommand(taxiC, "RRRRRDUUULLLLD"));

        //act
        dispatcher.dispatch(taxiCommands);

        //assert
        Approvals.verify(TaxiPrinter.plotTaxis(listOfTaxis));

        System.out.println("TaxiA Number of move commands issued: " + taxiCommands.get(0).command().length());
        System.out.println("TaxiB Number of move commands issued: " + taxiCommands.get(1).command().length());
        System.out.println("TaxiC Number of move commands issued: " + taxiCommands.get(2).command().length());

        System.out.println("TaxiA distance travelled: " + taxiA.cellometer());
        System.out.println("TaxiB distance travelled: " + taxiB.cellometer());
        System.out.println("TaxiC distance travelled: " + taxiC.cellometer());

    }





}
