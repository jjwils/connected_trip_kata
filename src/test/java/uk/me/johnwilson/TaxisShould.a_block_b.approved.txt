    0    1    2    3    4    5

0 [   ][ B ][   ][   ][   ][   ]

1 [   ][A  ][   ][   ][   ][   ]

2 [   ][   ][   ][   ][   ][   ]

3 [   ][   ][   ][   ][   ][   ]
